libshvproto-macros/src/lib.rs

```diff
@@ -80,11 +128,88 @@ pub fn derive_from_rpcvalue(item: TokenStream) -> TokenStream {
+                    if should {
+                        expected_types.extend(quote! {+ " " + stringify!(#dest_variant_type)});
+
+                        try_froms.extend(quote!{
+                            if let shvproto::Value::#dest_variant_type = value.value() {
+                                return Ok(<#struct_identifier>::#variant_ident #block);
+                            }
+                        });
+                    }
+                };
+                if let syn::Fields::Unnamed(variant_types) = &variant.fields {
+                    if variant_types.unnamed.len() != 1 {
+                        panic!("jde jenom jeden typ" );
+                    }
+                    match_arms.extend(quote!{
+                        #struct_identifier::#variant_ident(val) => RpcValue::from(val),
+                    });
+
+                    let source_variant_type = &variant_types.unnamed.first().expect("variant_type").ty;
+                    let deref_code = quote!((*x));
+                    let unbox_code = quote!((x.as_ref().clone()));
+
+                    add_type_matcher(is_type(source_variant_type, "i64"), quote!{Int(x)}, (deref_code).clone());
+                    add_type_matcher(is_type(source_variant_type, "u64"), quote!{UInt(x)}, (deref_code).clone());
+                    add_type_matcher(is_type(source_variant_type, "f64"), quote!{Double(x)}, (deref_code).clone());
+                    add_type_matcher(is_type(source_variant_type, "bool"), quote!{Bool(x)}, (deref_code).clone());
+                    add_type_matcher(is_type(source_variant_type, "DateTime"), quote!{DateTime(x)}, (deref_code).clone());
+                    add_type_matcher(is_type(source_variant_type, "Decimal"), quote!{Decimal(x)}, (deref_code).clone());
+                    add_type_matcher(is_type(source_variant_type, "String"), quote!{String(x)}, (unbox_code).clone());
+                    add_type_matcher(is_type(source_variant_type, "Blob"), quote!{Blob(x)}, (unbox_code).clone());
+                    add_type_matcher(is_type(source_variant_type, "List"), quote!{List(x)}, (unbox_code).clone());
+                    add_type_matcher(is_type(source_variant_type, "Map"), quote!{Map(x)}, (unbox_code).clone());
+                    add_type_matcher(is_type(source_variant_type, "IMap"), quote!{IMap(x)}, (unbox_code).clone());
+                }
+                if let syn::Fields::Unit = &variant.fields {
+                    match_arms.extend(quote!{
+                        #struct_identifier::#variant_ident => RpcValue::null(),
+                    });
+                    add_type_matcher(true, quote! {Null}, quote!());
+                }
+            }
+
+            quote!{
+                impl TryFrom<shvproto::RpcValue> for #struct_identifier {
+                    type Error = String;
+                    fn try_from(value: shvproto::RpcValue) -> Result<Self, Self::Error> {
+                        #try_froms
```

Comment:
If `try_froms` are match arms, this would be:
``` rust
match value.value() {
    #try_froms
    _ => Err("Couldn't deserialize ".to_owned() + stringify!(#struct_identifier) + ", expected types: " + #expected_types + ", got: " + value.type_name())
}

```